thoth/common/workflows.py

```diff
@@ -0,0 +1,234 @@
+# thoth-common
+# Copyright(C) 2019 Marek Cermak
+#
+# This program is free software: you can redistribute it and / or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Workflow management for Thoth."""
+
+import logging
+import json
+import random
+import re
+import requests
+import yaml
+
+from pathlib import Path
+
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Mapping
+from typing import Optional
+from typing import Union
+
+from attrdict import AttrDict
+
+from argo.workflows import client
+from argo.workflows import models
+
+from .exceptions import ConfigurationError
+from .exceptions import WorkflowError
+
+from .helpers import to_camel_case
+from .helpers import to_snake_case
+
+from .openshift import OpenShift
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Workflow(models.V1alpha1Workflow):  # type: ignore
+    """Argo Workflow instance.
+
+    This is a subclass of argo.workflows V1alpha1Workflow model
+    which provides a convenient set of methods to make workflow
+    managemend easier.
+    """
+
+    def __init__(
+        self,
+        api_version: str,
+        kind: str,
+        metadata: models.V1alpha1Metadata,
+        spec: models.V1alpha1WorkflowSpec,
+        status: Optional[models.V1alpha1WorkflowStatus] = None,
+    ):
+        """Initialize Workflow instance."""
+        super().__init__(
+            api_version=api_version, kind=kind, metadata=metadata, spec=spec, status=status or {},
+        )
+
+        if not status:
+            # fixup the status empty dict to prevent serialization issues
+            self.status = None
+
+        self.__validated = False
+
+    @property
+    def name(self) -> Union[str, None]:
+        """Get Workflow name."""
+        name: Union[str, None] = getattr(self.metadata, "name", None)
+        return name
+
+    @property
+    def id(self) -> str:
+        """Get Workflow ID."""
+        digest: str = abs(self.__hash__())
+        return f"workflow{'-' + self.name if self.name else ''}-{digest}"
```

Comment:
It might be a good idea to reuse digest creation from the openshift instance to unify digest creation (you can make it public):

https://github.com/thoth-station/common/blob/c21cd41a3a5f52c756225f34b7e7f4cc16f21bfb/thoth/common/openshift.py#L1137-L1139